File: wx-proxy/wx-proxy.py
#!/usr/bin/env python3
"""
YYB-Go-Enhanced wx/code 兼容代理
用 Python 3.11 stdlib 实现，不需要任何第三方包。

目的：让 smallfawn 青龙脚本零改动接入 YYB 作为微信协议后端。

脚本调用格式（smallcat 标准）：
  POST http://<host>:8787/wx/code
  headers: { auth: <token>, Content-Type: application/json }
  body: { appid: '<appid>', openid: '<openid>' }
  response: { status: true, data: { code: '...' } }

本代理转发到 YYB /wxapp/getCode，做格式转换：
  请求 {openid, appid} → YYB {ref: '<openid>', app_id: '<appid>'}
  响应 YYB {data.result.code} → {status:true, data:{code: '...'}}

启动：python3 wx-proxy.py
环境变量：
  PROXY_PORT   监听端口，默认 8787
  YYB_URL      YYB 服务地址，默认 http://127.0.0.1:9001
  AUTH_TOKEN   认证 token，默认 aa06c54e82f5439dc025e5f223b6466f
"""

import json
import os
import http.server
import urllib.request
import urllib.error


PORT = int(os.environ.get("PROXY_PORT", "8787"))
YYB_URL = os.environ.get("YYB_URL", "http://127.0.0.1:9001").rstrip("/")
AUTH_TOKEN = os.environ.get("AUTH_TOKEN", "aa06c54e82f5439dc025e5f223b6466f")


def call_yyb(openid, appid):
    """转发到 YYB /wxapp/getCode，返回 code 字符串或 None"""
    body = json.dumps({"ref": openid, "app_id": appid}).encode("utf-8")
    req = urllib.request.Request(
        f"{YYB_URL}/wxapp/getCode",
        data=body,
        headers={"Content-Type": "application/json"},
        method="POST",
    )
    try:
        with urllib.request.urlopen(req, timeout=15) as resp:
            data = json.loads(resp.read().decode("utf-8"))
    except urllib.error.HTTPError as e:
        return None, f"HTTP {e.code}: {e.read().decode('utf-8', errors='replace')}"
    except Exception as e:
        return None, f"request error: {e}"

    code = (
        data.get("data")
        or {}
    )
    if isinstance(code, dict):
        code = (code.get("result") or {}).get("code", "")
    return (code, data.get("msg", "")) if code else (None, data.get("msg", ""))


class Handler(http.server.BaseHTTPRequestHandler):
    def log_message(self, fmt, *args):
        # 减少日志，只记录错误
        if args and "502" in str(args):
            super().log_message(fmt, *args)

    def _send_json(self, status, obj):
        body = json.dumps(obj, ensure_ascii=False).encode("utf-8")
        self.send_response(status)
        self.send_header("Content-Type", "application/json")
        self.send_header("Content-Length", str(len(body)))
        self.end_headers()
        self.wfile.write(body)

    def do_POST(self):
        length = int(self.headers.get("Content-Length", "0"))
        body = self.rfile.read(length) if length > 0 else b""

        path = self.path.rstrip("/")

        # /wx/code — 核心兼容接口
        if path == "/wx/code":
            try:
                parsed = json.loads(body) if body else {}
            except Exception:
                parsed = {}

            openid = (parsed.get("openid") or "").strip()
            appid = (parsed.get("appid") or parsed.get("app_id") or "").strip()

            if not openid or not appid:
                self._send_json(400, {"status": False, "message": "openid and appid required"})
                return

            code, err = call_yyb(openid, appid)
            if code:
                self._send_json(200, {"status": True, "message": "获取成功", "data": {"code": code}})
            else:
                self._send_json(502, {"status": False, "message": f"获取 code 失败: {err}"})
            return

        # /wx/getuserinfo — 部分脚本用此端点
        if path == "/wx/getuserinfo":
            try:
                parsed = json.loads(body) if body else {}
            except Exception:
                parsed = {}
            openid = (parsed.get("openid") or "").strip()
            appid = (parsed.get("appid") or parsed.get("app_id") or "").strip()
            if not openid or not appid:
                self._send_json(400, {"status": False, "message": "openid and appid required"})
                return
            code, err = call_yyb(openid, appid)
            if code:
                self._send_json(200, {"status": True, "data": {"code": code, "encryptedData": "placeholder", "iv": "placeholder"}})
            else:
                self._send_json(502, {"status": False, "message": f"获取 code 失败: {err}"})
            return
        # 未识别 POST
        self._send_json(404, {"status": False, "message": "not found: " + path})

    def do_GET(self):
        path = self.path.rstrip("/")

        # /health
        if path == "/health" or path == "":
            self._send_json(200, {"status": True, "data": {"ok": True}})
            return

        # /credits/balance — YYB 无积分机制，始终返回充足
        if path == "/credits/balance":
            self._send_json(200, {"status": True, "data": {"balance": 99999}})
            return

        self._send_json(404, {"status": False, "message": "not found: " + path})


if __name__ == "__main__":
    server = http.server.HTTPServer(("0.0.0.0", PORT), Handler)
    print(f"[wx-proxy] listening on 0.0.0.0:{PORT}")
    print(f"[wx-proxy] forwarding to {YYB_URL}/wxapp/getCode")
    print(f"[wx-proxy] auth token: {AUTH_TOKEN}")
    server.serve_forever()

File: wx-proxy/wx-proxy.js
/**
 * YYB-Go-Enhanced wx/code 兼容代理
 *
 * 目的：让 smallfawn 青龙脚本零改动接入 YYB 作为微信协议后端。
 * 脚本调用格式（smallcat 标准）：
 *   POST http://<host>:8787/wx/code
 *   headers: { auth: <token>, Content-Type: application/json }
 *   body: { appid: '<appid>', openid: '<openid>' }
 *   response: { status: true, data: { code: '...' } }
 *
 * 本代理转发到 YYB /wxapp/getCode，做格式转换：
 *   请求 {openid, appid} → YYB {ref: '<openid>', app_id: '<appid>'}
 *   响应 YYB {data.result.code} → {status:true, data:{code: '...'}}
 *
 * 环境：node >= 16（无需 npm install，纯 stdlib 写法用 fetch）
 *
 * 启动：node wx-proxy.js
 * 环境变量：
 *   PROXY_PORT       监听端口，默认 8787
 *   YYB_URL          YYB 服务地址，默认 http://127.0.0.1:9001
 *   AUTH_TOKEN       认证 token，默认 aa06c54e82f5439dc025e5f223b6466f
 */

const http = require("http");

const PORT = parseInt(process.env.PROXY_PORT, 10) || 8787;
const YYB_URL = (process.env.YYB_URL || "http://127.0.0.1:9001").replace(/\/$/, "");
const AUTH_TOKEN = process.env.AUTH_TOKEN || "aa06c54e82f5439dc025e5f223b6466f";

function parseUrl(url) {
  // 简易解析
  const m = url.match(/^https?:\/\/([^/]+)(\/.*)/);
  return m ? { host: m[1], path: m[2] } : null;
}

function request(opts) {
  return new Promise((resolve, reject) => {
    const parsed = parseUrl(opts.url);
    const method = opts.method || "POST";
    const body = opts.data ? JSON.stringify(opts.data) : null;

    const req = http.request(
      {
        hostname: parsed.host,
        port: 80,
        path: parsed.path,
        method,
        headers: {
          "Content-Type": "application/json",
          "Content-Length": body ? Buffer.byteLength(body) : 0,
          ...(opts.headers || {}),
        },
        timeout: 15000,
      },
      (res) => {
        let chunks = "";
        res.on("data", (c) => (chunks += c));
        res.on("end", () => {
          let data = null;
          try {
            data = JSON.parse(chunks);
          } catch (e) {
            data = { _raw: chunks };
          }
          resolve({ status: res.statusCode, data, headers: res.headers || {} });
        });
      }
    );
    req.on("error", reject);
    req.on("timeout", () => {
      req.destroy();
      reject(new Error("timeout"));
    });
    if (body) req.write(body);
    req.end();
  });
}

const server = http.createServer(async (req, res) => {
  let body = "";
  req.on("data", (c) => (body += c));
  req.on("end", async () => {
    const path = req.url || "";

    // /wx/code — 主要兼容接口
    if (path === "/wx/code" || path === "/wx/code/") {
      try {
        let parsed = {};
        try {
          parsed = JSON.parse(body) || {};
        } catch (e) {}

        const openid = parsed.openid || "";
        const appid = parsed.appid || parsed.app_id || "";
        if (!openid || !appid) {
          res.writeHead(400, { "Content-Type": "application/json" });
          res.end(JSON.stringify({ status: false, message: "openid and appid required" }));
          return;
        }

        // 转发到 YYB
        const { status, data } = await request({
          method: "POST",
          url: `${YYB_URL}/wxapp/getCode`,
          data: { ref: openid, app_id: appid },
        });

        // YYB 响应格式转换
        const code =
          (data && data.data && data.data.result && data.data.result.code) || "";

        if (status === 200 && code) {
          res.writeHead(200, { "Content-Type": "application/json" });
          res.end(JSON.stringify({ status: true, message: "获取成功", data: { code } }));
        } else {
          const errMsg =
            (data && data.msg) || `HTTP ${status}`;
          res.writeHead(502, { "Content-Type": "application/json" });
          res.end(
            JSON.stringify({
              status: false,
              message: `获取 code 失败: ${errMsg}`,
              data: data || {},
            })
          );
        }
      } catch (err) {
        res.writeHead(502, { "Content-Type": "application/json" });
        res.end(JSON.stringify({ status: false, message: err.message, data: null }));
      }
      return;
    }

    // /api/client/login — 兼容 smallcat 登录接口（用于脚本初始化）
    if (path.startsWith("/api/client/login")) {
      res.writeHead(200, { "Content-Type": "application/json" });
      res.end(
        JSON.stringify({
          status: true,
          data: {
            auth: AUTH_TOKEN,
            nickname: "YYB-Proxy",
            version: "1.1.2",
          },
        })
      );
      return;
    }

    // /health — 健康检查
    if (path === "/health" || path === "/") {
      res.writeHead(200, { "Content-Type": "application/json" });
      res.end(JSON.stringify({ status: true, data: { ok: true } }));
      return;
    }

    // /credits/balance — 兼容 smallcat 积分查询（YYB 无积分，始终返回充足）
    if (path === "/credits/balance") {
      res.writeHead(200, { "Content-Type": "application/json" });
      res.end(JSON.stringify({ status: true, data: { balance: 99999 } }));
      return;
    }

    // 未识别路径
    res.writeHead(404, { "Content-Type": "application/json" });
    res.end(JSON.stringify({ status: false, message: "not found: " + path }));
  });
});

server.listen(PORT, "0.0.0.0", () => {
  console.log(`[wx-proxy] listening on 0.0.0.0:${PORT}`);
  console.log(`[wx-proxy] forwarding to ${YYB_URL}/wxapp/getCode`);
});
